### DISCUSSION TITLE
Getting admin_centre node in process_relation

### DISCUSSION THREAD

**Discussion by zdila** [osm2pgsql-dev/osm2pgsql#1763] [Q&A] [answered]
Hello,

I need node with name and geometry (point) in `process_relation` function. My current approach is following:

```lua
local node_table = {};

function osm2pgsql.process_node(object)
  if object.tags.place then -- consider only places not to fill-up whole RAM
    node_table[object.id] = { point = object:as_point(), name = object.tags.name }
  end
end

function osm2pgsql.process_relation(object)
  local type = object.tags.type;

  for _, member in ipairs(object.members) do
    if member.role == 'admin_centre' and member.type == 'n' then
      local node = node_table[member.ref]

      // now I can use `node.name` and `node.point`
    end
  end
end
```

Is it correct? I mean it works but it requires storing potential nodes to `node_table` which eats RAM. I would expect to be able to access nodes (and ways) including geometry in `process_relation` via some API but I couldn't find it.

Maybe another approach could be to store references to admin_centre nodes in `select_relation_members` and then look them up in `process_node` to save some memory?

**Comment by joto** [comment, marked as answer]
You can get a collection of all member geometries of a relation with `object:as_geometrycollection()` and then get the nth geometry with `geometry_n()`. So something like this should work:

```
function osm2pgsql.process_relation(object)
  local geom = object.as_geometrycollection()
  for i, member in ipairs(object.members) do
    if member.role == 'admin_centre' and member.type == 'n' then
      local point_geom = geom.geometry_n(i)
      ... 
    end
  end
end
```

But that only gives you the geometry, not the name.

If you don't want to store *all* place points, but only those actually mentioned in a relation, you can look at [two-stage processing](https://osm2pgsql.org/doc/manual.html#stages). I's a bit tricky to get right though.

You can also just take advantage of the database you have. Add those place nodes to a database table, create another table from the relation members and join them afterwards.

**Reply by zdila** [reply]
Thanks @joto for your valuable hints!
